plugins/junit/src/com/intellij/execution/junit/codeInsight/JUnit5MalformedOrderedTestInspection.kt

```diff
@@ -0,0 +1,165 @@
+      }
+
+      /**
+       * @return `null` when problem has been found, otherwise no problem decided returning instance of [@PsiAnnotation]
+       * on which decision could be made
+       */
+      private fun detectProblemPsiClassWithoutTestMethodOrderAnnotation(
+        classWithTestMethodOrderer: PsiClass,
+        problemOnPsiElement: PsiElement,
+        problemDescription: String
+      ): PsiAnnotation? {
+        val testMethodOrderAnnotation = AnnotationUtil.findAnnotationInHierarchy(classWithTestMethodOrderer, setOf(
+          JUnitCommonClassNames.ORG_JUNIT_JUPITER_API_TEST_METHOD_ORDER))
+        if (testMethodOrderAnnotation == null) {
+          holder.registerProblem(
+            problemOnPsiElement,
+            problemDescription,
+            object : AddAnnotationFix(
+              JUnitCommonClassNames.ORG_JUNIT_JUPITER_API_TEST_METHOD_ORDER,
+              classWithTestMethodOrderer,
+              createAnnotationValues(classWithTestMethodOrderer)) {
+              override fun getFamilyName(): String = "Add @TestMethodOrder with @Order support"
+            })
+          return null
+        }
+        else {
+          return testMethodOrderAnnotation
+        }
+      }
+
+      private fun detectProblemWithTestMethodOrderAnnotationValue(
+        testMethodOrderAnnotation: PsiAnnotation,
+        placeOnPsiClass: PsiClass,
+        problemOnPsiElement: PsiElement,
+        problemDescription: String
+      ) {
+        val value = testMethodOrderAnnotation.findAttributeValue("value")
+                      as PsiClassObjectAccessExpression? ?: return
+
+        val valuePsiClass = PsiTypesUtil.getPsiClass(value.operand.type) ?: return
+        if (PsiClassImplUtil.isClassEquivalentTo(valuePsiClass, psiClassOrderAnnotationOrderer)) {
+          // annotation value is known correct MethodOrderer implementation
+          return
+        }
+
+        val packageName = PsiUtil.getPackageName(valuePsiClass) ?: return
+        if (packageName.startsWith("org.junit.jupiter.api")) {
+          // other JUnit5's MethodOrderer implementations ignore @Order annotation
+          holder.registerProblem(
+            problemOnPsiElement,
+            problemDescription,
+            object : RemoveAnnotationQuickFix(testMethodOrderAnnotation, placeOnPsiClass) {
+              // 1st LocalQuickFix's familyName is used to show to the user
+              override fun getFamilyName(): String {
+                return "Replace @TestMethodOrder value to support @Order"
+              }
+            },
+            AddAnnotationFix(
+              JUnitCommonClassNames.ORG_JUNIT_JUPITER_API_TEST_METHOD_ORDER,
+              placeOnPsiClass,
+              createAnnotationValues(placeOnPsiClass)))
+        }
+        // but custom implementation of MethodOrderer may take @Order annotation into account (cannot decide)
+      }
+
+      private fun createAnnotationValues(context: PsiClass): Array<out PsiNameValuePair>? {
```

Comment:
I am also not sure about this creation of `PsiNameValuePair` instances, I mean that creation of artificial annotation from text and then getting list out of it.